File: CodeWars/Java/6Kyu/HammingCode.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.stream.*;

public class HammingCode {

    public static String encode(String text) {
        List<String> letters = new ArrayList<String>();

        char[] chars = text.toCharArray();

        for (char ch: chars){
            int asciiNumber = (int) ch;

            String binary = Integer.toBinaryString(asciiNumber);

            while(binary.length()<8){
                binary = "0" + binary;
            }

            char[] numbers = binary.toCharArray();

            for (char number: numbers){
                letters.addAll(Collections.nCopies(3, Character.toString(number)));
            }
        }

        return String.join("", letters);
    }

    public static String decode(String text) {
        String letters = "";
        
        for (int i = 0; i < text.length(); i += 3) {
            String threeCharacters = text.substring(i, i + 3);
            
            List<String> threeCharactersList = Arrays.asList(threeCharacters.split(""));

            Map<String, Long> occurrences = 
                threeCharactersList.stream().collect(Collectors.groupingBy(w -> w, Collectors.counting()));
            
            System.out.println("oxe");
            System.out.println(occurrences);

            if (occurrences.containsKey("1")){
                letters += '1';
            }
            else{
                letters += '0';
            }
        }

        List<String> asciLetters = new ArrayList<String>();

        for (int i = 0; i < letters.length(); i += 8) {
            String asciLetter = letters.substring(i, i + 8);
            
            asciLetters.add(Character.toString((char) Integer.parseInt(asciLetter,2)));
        }

        return String.join("", asciLetters);
    }

  public static void main(String[] args) {
    // System.out.println(encode("hey"));
    System.out.println(decode("000111111000111000000000000111111000000111000111000111111111111000000111"));
  }
}

File: CodeWars/JavaScript/6Kyu/HammingCode.js
// URL: https://www.codewars.com/kata/5ef9ca8b76be6d001d5e1c3e/train/javascript

function encode(text) {
    let letters = [];

    for (character of text){
        byte = character.charCodeAt(0).toString(2);
        
        while(byte.length<8){
            byte = '0' + byte;
        }

        for (bit of byte){
            letters.push(bit.repeat(3));
        }
    }

    return letters.join('');
}

function decode(bits) {
    let letters = '';

    for (let i = 0, charsLength = bits.length; i < charsLength; i += 3) {
        const threeCharacters = bits.substring(i, i + 3);
        
        if (threeCharacters.replace(/[^1]/g, "").length > threeCharacters.replace(/[^0]/g, "").length){
            letters += '1';
        }
        else{
            letters += '0';
        }
    }
    
    let asciLetters = [];

    for (let i = 0, charsLength = letters.length; i < charsLength; i += 8) {
        const asciLetter = letters.substring(i, i + 8);
        
        asciLetters.push(String.fromCodePoint(parseInt(asciLetter, 2)));
    }

    return asciLetters.join('');
}

console.log(decode("100111111000111001000010000111111000000111001111000111110110111000010111"));